src/Sentry/Internal/MainSentryEventProcessor.cs

```diff
@@ -96,19 +97,36 @@ public SentryEvent Process(SentryEvent @event)
                 }
             }
 
-            if (_options.ReportAssemblies)
+            if (_options.ReportAssembliesMode != ReportAssembliesMode.None)
             {
                 foreach (var assembly in AppDomain.CurrentDomain.GetAssemblies())
                 {
-                    if (assembly.IsDynamic)
+                    if (assembly is null ||
+                        assembly.IsDynamic)
                     {
                         continue;
                     }
 
                     var asmName = assembly.GetName();
-                    if (asmName.Name is not null && asmName.Version is not null)
+                    if (asmName.Name is not null)
                     {
-                        @event.Modules[asmName.Name] = asmName.Version.ToString();
+                        var asmVersion = string.Empty;
+
+                        switch (_options.ReportAssembliesMode)
+                        {
+                            case ReportAssembliesMode.Version:
+                                asmVersion = asmName.Version?.ToString() ?? string.Empty;
+                                break;
+
+                            case ReportAssembliesMode.InformationalVersion:
+                                asmVersion = assembly.GetNameAndVersion().Version ?? string.Empty;
+                                break;
+
+                            default:
+                                throw new ArgumentOutOfRangeException(nameof(_options.ReportAssembliesMode), $"Report assemblies mode '{_options.ReportAssembliesMode}' is not yet supported");
+                        }
+
+                        @event.Modules[asmName.Name] = asmVersion;
```

Comment:
To keep the behavior like before I think we need to make sure we're not adding it if we didn't resolve the version:

```suggestion
                        if (!string.IsNullOrWhiteSpace(asmVersion))
                        {
                            @event.Modules[asmName.Name] = asmVersion;
                        }
```